docs/api/js/utils.md

```diff
@@ -1,68 +1,772 @@
+        ...ethers.utils.arrayify(L1_MESSENGER_ADDRESS),
+        ...ethers.utils.zeroPad(sender, 32),
+        ...ethers.utils.arrayify(ethers.utils.keccak256(msg))
+    ]);
+
+    return ethers.utils.keccak256(encodedMsg);
+}
+```
+
+### `getSignature`
+
+Returns the signature of any transaction as a `Uint8Array`.
+
+#### Inputs
+
+- `transaction`: transaction address ?? as ??.
+- `ethSignature?`: transaction signature ?? as `EthereumSignature` object (optional).
+
+```ts
+function getSignature(transaction: any, ethSignature?: EthereumSignature): Uint8Array {
+    if (transaction?.customData?.customSignature && transaction.customData.customSignature.length) {
+        return ethers.utils.arrayify(transaction.customData.customSignature);
+    }
+
+    if (!ethSignature) {
+        throw new Error('No signature provided');
+    }
+
+    const r = ethers.utils.zeroPad(ethers.utils.arrayify(ethSignature.r), 32);
+    const s = ethers.utils.zeroPad(ethers.utils.arrayify(ethSignature.s), 32);
+    const v = ethSignature.v;
+
+    return new Uint8Array([...r, ...s, v]);
+}
+```
+
+### `hashBytecode`
+
+Returns the hash of given bytecode.
+
+#### Inputs
+
+- `bytecode`: ethers `BytesLike` object.
+
+```ts
+export function hashBytecode(bytecode: ethers.BytesLike): Uint8Array {
+    // For getting the consistent length we first convert the bytecode to UInt8Array
+    const bytecodeAsArray = ethers.utils.arrayify(bytecode);
+
+    if (bytecodeAsArray.length % 32 != 0) {
+        throw new Error('The bytecode length in bytes must be divisible by 32');
+    }
+
+    if (bytecodeAsArray.length > MAX_BYTECODE_LEN_BYTES) {
+        throw new Error(`Bytecode can not be longer than ${MAX_BYTECODE_LEN_BYTES} bytes`);
+    }
+
+    const hashStr = ethers.utils.sha256(bytecodeAsArray);
+    const hash = ethers.utils.arrayify(hashStr);
+
+    // Note that the length of the bytecode
+    // should be provided in 32-byte words.
+    const bytecodeLengthInWords = bytecodeAsArray.length / 32;
+    if (bytecodeLengthInWords % 2 == 0) {
+        throw new Error('Bytecode length in 32-byte words must be odd');
+    }
+
+    const bytecodeLength = ethers.utils.arrayify(bytecodeLengthInWords);
+
+    // The bytecode should always take the first 2 bytes of the bytecode hash,
+    // so we pad it from the left in case the length is smaller than 2 bytes.
+    const bytecodeLengthPadded = ethers.utils.zeroPad(bytecodeLength, 2);
+
+    const codeHashVersion = new Uint8Array([1, 0]);
+    hash.set(codeHashVersion, 0);
+    hash.set(bytecodeLengthPadded, 2);
+
+    return hash;
+}
+```
+
+### `isECDSASignatureCorrect`
+
+Like similar functionality in `ethers.js` but with added try/catch facility. The function returns true if the validation process succeeds.
+
+Called from [`isSignatureCorrect`](#isSignatureCorrect) for non-contract account addresses. 
+
+#### Inputs
+
+- `address`: sender address ?? as string.
```

Comment:
the address which should've signed the `msgHash`

Reply:
done